### PR TITLE
Add registerClass and related functions to handle derived class. fixes #55

### LINKED ISSUE #54: Don't know how to construct class type 'S' with argument types '()'
Seems like msgpack-d requires empty constructor for classes.
Can we relax this?
Classes could be allocated without calling 'new' constructor to enable this.

### LINKED ISSUE #55: register derived class without a custom handler
Registering a derived class should not require a custom handler, it should allow the following:

//custom code
registerPackHandler!(Subclass, cPackHandler);

//MISSING: default 'pack' code applied to derived Subclass when given a base class reference
registerPackHandler!(Subclass);

(likewise with unpack)

see how it's done with orange:

Serializer.register!(Sub);
https://github.com/jacob-carlborg/orange/blob/d5b2e0127c67f50bd885ee43a7dd61dd418b1661/tests/BaseClass.d#L58

### PR DESCRIPTION
**Pull request by repeatedly**
This changes null argument behaviour for object deserialization.
If you pass null object, Unpacker creates new object without calling constructor.

This also fixes #54.

### CODE REVIEW

**Comment by repeatedly** [comment]
@timotheecour Could you test this patch?

File: src/msgpack.d [hunk 1]
```diff
@@ -2760,6 +2748,53 @@ struct Unpacker
     }
 
 
+    void unpackObject(T)(ref T object) if (is(Unqual!T == class))
```
Comment (timotheecour) [inline src/msgpack.d]: s/is(Unqual!T == class))/is(T==class)
Unqual seems un-necessary:
assert(is(Unqual!(const(Object)) == class)); // passes

**Comment by timotheecour** [comment]
Thanks! that works indeeed!
I wrote a few comments on that pull request

File: src/msgpack.d (hunk 1 above)
Comment (repeatedly) [inline src/msgpack.d]: IIRC, Unqual is needed, but It could be removed in recent compilers.

**Pull request merged**